app/services/platform-apps/container-manager.ts

```diff
@@ -0,0 +1,360 @@
+import { ILoadedApp, EAppPageSlot } from '.';
+import electron from 'electron';
+import { trim, compact } from 'lodash';
+import { Inject } from 'util/injector';
+import { UserService } from 'services/user';
+import url from 'url';
+import path from 'path';
+import { PlatformAppsApi } from './api';
+import { lazyModule } from 'util/lazy-module';
+import { GuestApiService } from 'services/guest-api';
+import { BehaviorSubject } from 'rxjs';
+import { IBrowserViewTransform } from './api/modules/module';
+
+interface IContainerInfo {
+  appId: string;
+  slot: EAppPageSlot;
+  persistent: boolean;
+  container: electron.BrowserView;
+  transform: BehaviorSubject<IBrowserViewTransform>;
+}
+
+/**
+ * Manages the life cycle of application containers.  Application
+ * containers are restricted/sandboxed pages that can be mounted
+ * into any window.  These are implemented with electron BrowserViews.
+ */
+export class PlatformContainerManager {
+  containers: IContainerInfo[] = [];
+
+  @Inject() private userService: UserService;
+  @Inject() private guestApiService: GuestApiService;
+
+  @lazyModule(PlatformAppsApi) private apiManager: PlatformAppsApi;
+
+  /**
+   * Registers an app with the container service.
+   * Any persistent pages will immediately be spun
+   * up in the background.
+   * @param app The app to register
+   */
+  registerApp(app: ILoadedApp) {
+    // Make sure this app isn't already registered
+    this.unregisterApp(app);
```

Comment:
Do we have a more reliable way of checking this? I could be concerned about side-effects that `unregisterApp` might bring. For instance, when listening for app load/unload events in platform.

EDIT: Apparently none. Still, a bit hard to reason about.

Reply:
Yeah all unregister app does is shut down any running containers.  This is an additional layer of security to make sure that we aren't already running containers for this app before we start up new ones.  Perhaps this method doesn't have the best name, but I think the logic is sound.